### PR TITLE
Adjustment to BDEF Object Handler

### PR DESCRIPTION
**Pull request by ThomasPloski**
Adjustment to BDEF Object Handler regarding exchange file format
issue #3351

### CODE REVIEW

**Comment by larshp** [comment]
@christianguenter2 any comments?

**Comment by christianguenter2** [comment]
@ThomasPloski can you update the BDEF test case here?
https://github.com/abapGit-tests/BDEF

File: src/objects/zcl_abapgit_object_bdef.clas.abap [hunk 1]
```diff
@@ -204,72 +274,101 @@ CLASS zcl_abapgit_object_bdef IMPLEMENTATION.
...
+            WHEN '1'. "if_wb_adt_plugin_resource_co=>co_sfs_res_category_atomic.
+              CALL METHOD lo_wb_object_operator->('IF_WB_OBJECT_OPERATOR~CREATE')
+                EXPORTING
+                  io_object_data    = lo_object_data
+                  data_selection    = 'AL' " if_wb_object_data_selection_co=>c_all_data
+                  version           = 'I'
+                  package           = iv_package
+                  transport_request = lv_transport_request.
+            WHEN '2'. "if_wb_adt_plugin_resource_co=>co_sfs_res_category_compound_s.
+              CALL METHOD lo_wb_object_operator->('CREATE')
```
Comment (christianguenter2) [inline src/objects/zcl_abapgit_object_bdef.clas.abap:327]: Doesn't work for me on S/4HANA 1909 SP02 

image

CREATE method doesn't exist in CL_WB_OBJECT_OPERATOR.
image

**Review by christianguenter2 — CHANGES_REQUESTED** [review]
@ThomasPloski import doesn't work for me on S/4HANA 1909 SP02. See comment.

**Comment by ThomasPloski** [comment]
Hi @christianguenter2, I have adapted the method call to the corresponding interface method.

File: src/objects/zcl_abapgit_object_bdef.clas.abap [hunk 2]
```diff
@@ -204,72 +274,101 @@ CLASS zcl_abapgit_object_bdef IMPLEMENTATION.
...
+                  package           = iv_package
+                  transport_request = lv_transport_request.
+            WHEN '2'. "if_wb_adt_plugin_resource_co=>co_sfs_res_category_compound_s.
+              CALL METHOD lo_wb_object_operator->('IF_WB_OBJECT_OPERATOR~CREATE')
+                EXPORTING
+                  io_object_data    = lo_object_data
+                  data_selection    = 'P' " if_wb_object_data_selection_co=>c_properties
+                  version           = 'I'
+                  package           = iv_package
+                  transport_request = lv_transport_request.
```
Comment (christianguenter2) [inline src/objects/zcl_abapgit_object_bdef.clas.abap:333]: @ThomasPloski Next error
image

image

**Review by christianguenter2 — CHANGES_REQUESTED** [review]

File: src/objects/zcl_abapgit_object_bdef.clas.abap (hunk 2 above)
Comment (larshp) [reply src/objects/zcl_abapgit_object_bdef.clas.abap:333]: @ThomasPloski

Comment (larshp) [reply src/objects/zcl_abapgit_object_bdef.clas.abap:333]: @ThomasPloski

**Comment by larshp** [comment]
@ThomasPloski old method `fill_metadata_from_db` can be deleted?

File: src/objects/zcl_abapgit_object_bdef.clas.abap (hunk 2 above)
Comment (larshp) [reply src/objects/zcl_abapgit_object_bdef.clas.abap:484]: I guess this was fixed in the commits below? @christianguenter2 can you try again?

**Comment by larshp** [comment]
@christianguenter2 do you have time to take another look?

File: src/objects/zcl_abapgit_object_bdef.clas.abap [hunk 3]
```diff
@@ -151,6 +216,162 @@ CLASS zcl_abapgit_object_bdef IMPLEMENTATION.
...
+    ASSERT sy-subrc = 0.
+
+    <lv_source> = mo_files->read_string( 'asbdef' ).
+
+    CREATE OBJECT ro_object_data TYPE ('CL_BLUE_SOURCE_OBJECT_DATA').
+
+    CALL METHOD ro_object_data->('IF_WB_OBJECT_DATA_MODEL~SET_DATA')
+      EXPORTING
+        p_data = <lg_data>.
+
```
Comment (christianguenter2) [inline src/objects/zcl_abapgit_object_bdef.clas.abap:261]: Error when pulling demo repo into S4/HANA 1909 SP02
image

**Review by christianguenter2 — CHANGES_REQUESTED** [review]
@ThomasPloski

**Comment by christianguenter2** [comment]
@ThomasPloski 
image

File: src/objects/zcl_abapgit_object_bdef.clas.abap [hunk 4]
```diff
@@ -485,4 +538,114 @@ CLASS zcl_abapgit_object_bdef IMPLEMENTATION.
...
+    ASSIGN COMPONENT 'CONTENT-SOURCE' OF STRUCTURE <lg_data> TO <lv_source>.
+    ASSERT sy-subrc = 0.
+
+    <lv_source> = mo_files->read_string( 'asbdef' ).
+
+    CREATE OBJECT ro_object_data TYPE ('CL_BLUE_SOURCE_OBJECT_DATA').
+
+    CALL METHOD ro_object_data->('IF_WB_OBJECT_DATA_MODEL~SET_DATA')
+      EXPORTING
+        p_data = <lg_data>.
```
Comment (christianguenter2) [inline src/objects/zcl_abapgit_object_bdef.clas.abap:583]: @ThomasPloski 
still get this one
image

**Review by christianguenter2 — CHANGES_REQUESTED** [review]
@ThomasPloski  see comment

**Comment by ThomasPloski** [comment]
Hi @christianguenter2 & @larshp,
Currently I have some issues with activating BDEF object. In case of I use the activation_old logic the activation proceed without issues.

**Comment by mbtools** [comment]
BDEF is not a DDIC type. It should use the following settings

`rs_metadata-ddic = abap_false.`  (or remove it)
`    APPEND zif_abapgit_object=>gc_step_id-abap TO rt_steps.`

Then it will default to `activation_old`.

**Comment by christianguenter2** [comment]
@ThomasPloski Is it ready for re-review?
I still get this error
image

**Comment by ThomasPloski** [comment]
> @ThomasPloski Is it ready for re-review?
> I still get this error
> image

Would it be possible to send me your repo which you use to test this changes? With my BDEF test objects I do not get this issue you have mentioned.

**Comment by christianguenter2** [comment]
@ThomasPloski 
Sure, it the official abapGit test repo for BDEF 
https://github.com/abapGit-tests/BDEF.git

**Comment by larshp** [comment]
@ThomasPloski

**Comment by christianguenter2** [comment]
@ThomasPloski

**Comment by larshp** [comment]
@ThomasPloski

**Comment by ThomasPloski** [comment]
@christianguenter2: I have changed the affected reference type of the BDEF create API. Now I am able import your BDEF example. Would you please double check and put changes to master? Thanks.

**Review by christianguenter2 — APPROVED** [review]
Tested. LGTM

**Comment by larshp** [comment]
thanks 👍

**Pull request merged**